### PR TITLE
[MySQL] Treat decimal as numeric-string

### LINKED ISSUE #428: PHP 8.1, PDO, MySQL, Decimal column
Hello,

`decimal` type is returned as `string` in PHP, but phpstan-dba thinks it's a float :cry:.

This is, of course, without stringified fetches.

(unless I miss something :scream:)

### PR DESCRIPTION
**Pull request by marmichalski**
Closes #428.

### CODE REVIEW

File: src/TypeMapping/MysqlTypeMapper.php [hunk 1]
```diff
@@ -117,13 +117,17 @@ public function mapToPHPStanType(string $mysqlType, array $mysqlFlags, int $leng
...
         // floats are detected as numerics in mysqli
-        if (\in_array(strtoupper($mysqlType), ['DOUBLE', 'NEWDECIMAL', 'REAL'], true)) {
+        if (\in_array(strtoupper($mysqlType), ['DOUBLE', 'REAL'], true)) {
             $phpstanType = new FloatType();
         }
 
         // fallbacks
         if (null === $phpstanType) {
             switch (strtoupper($mysqlType)) {
+                case 'DECIMAL':
```
Comment (marmichalski) [inline src/TypeMapping/MysqlTypeMapper.php:127]: (this is probably redundant)

File: tests/default/data/pdo-mysql.php [hunk 2]
```diff
@@ -18,14 +18,14 @@ public function execute(PDO $pdo)
...
     }
 
     public function aggregateFunctions(PDO $pdo)
     {
         $query = 'SELECT MAX(adaid), MIN(adaid), COUNT(adaid), AVG(adaid) FROM ada WHERE adaid = 1';
         $stmt = $pdo->query($query, PDO::FETCH_ASSOC);
-        assertType('PDOStatement<array{MAX(adaid): int<-32768, 32767>|null, MIN(adaid): int<-32768, 32767>|null, COUNT(adaid): int, AVG(adaid): float|null}>', $stmt);
+        assertType('PDOStatement<array{MAX(adaid): int<-32768, 32767>|null, MIN(adaid): int<-32768, 32767>|null, COUNT(adaid): int, AVG(adaid): numeric-string|null}>', $stmt);
```
Comment (marmichalski) [inline tests/default/data/pdo-mysql.php:28]: Tested and result of AVG is also a `string`.

**Review by staabm — APPROVED** [review]
Looks great thx. Did you verify the results on mysqli and pdo+mysql?

File: src/TypeMapping/MysqlTypeMapper.php [hunk 3]
```diff
@@ -117,13 +117,17 @@ public function mapToPHPStanType(string $mysqlType, array $mysqlFlags, int $leng
...
+                case 'NEWDECIMAL':
+                    $phpstanType = new AccessoryNumericStringType();
```
Comment (staabm) [inline src/TypeMapping/MysqlTypeMapper.php:129]: See line 161 how a numeric string is built. An accessory type cannot stand on irs own

Comment (marmichalski) [reply src/TypeMapping/MysqlTypeMapper.php:129]: Yea, I've seen it, but instantiated it just like this and it seems to work ? 🙈

**Comment by marmichalski** [comment]
> Looks great thx. Did you verify the results on mysqli and pdo+mysql?

I have tested only with pdo. I can do some more testing before we merge this

File: src/TypeMapping/MysqlTypeMapper.php (hunk 3 above)
Comment (staabm) [reply src/TypeMapping/MysqlTypeMapper.php:129]: It works for the tests but will not work in the bigger picture when the type needs to be combined with other logic within phpstan

**Comment by marmichalski** [comment]
> > Looks great thx. Did you verify the results on mysqli and pdo+mysql?
> 
> I have tested only with pdo. I can do some more testing before we merge this

<details>
  <summary>
    best code
  </summary>

```
// pdo5 -> pdo + mysql:5.7, pdo8 -> pdo + mysql:8.0; mysqli is mysqli
var_dump(
    $pdo5->query('SELECT eadavk FROM ak')->fetch(\PDO::FETCH_ASSOC),
    $pdo8->query('SELECT eadavk FROM ak')->fetch(\PDO::FETCH_ASSOC),

    $mysqli5->query('SELECT eadavk FROM ak')->fetch_row(),
    $mysqli8->query('SELECT eadavk FROM ak')->fetch_row(),
);
```
</details>

```
array(1) {
  ["eadavk"]=>
  string(6) "154.23"
}
array(1) {
  ["eadavk"]=>
  string(6) "154.23"
}
array(1) {
  [0]=>
  string(6) "154.23"
}
array(1) {
  [0]=>
  string(6) "154.23"
}
```

**Comment by staabm** [comment]
Awesome thx

**Pull request merged**